src/ray/core_worker/transport/direct_actor_task_submitter.cc

```diff
@@ -80,32 +83,39 @@ Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(TaskSpecification task_spe
       // this sequence number.
       send_pos = task_spec.ActorCounter();
       RAY_CHECK(queue->second.actor_submit_queue->Emplace(send_pos, task_spec));
+      queue->second.cur_pending_calls++;
       task_queued = true;
     }
   }
 
   if (task_queued) {
-    // We must release the lock before resolving the task dependencies since
-    // the callback may get called in the same call stack.
-    resolver_.ResolveDependencies(task_spec, [this, send_pos, actor_id](Status status) {
-      absl::MutexLock lock(&mu_);
-      auto queue = client_queues_.find(actor_id);
-      RAY_CHECK(queue != client_queues_.end());
-      auto &actor_submit_queue = queue->second.actor_submit_queue;
-      // Only dispatch tasks if the submitted task is still queued. The task
-      // may have been dequeued if the actor has since failed.
-      if (actor_submit_queue->Contains(send_pos)) {
-        if (status.ok()) {
-          actor_submit_queue->MarkDependencyResolved(send_pos);
-          SendPendingTasks(actor_id);
-        } else {
-          auto task_id = actor_submit_queue->Get(send_pos).first.TaskId();
-          actor_submit_queue->MarkDependencyFailed(send_pos);
-          task_finisher_.FailOrRetryPendingTask(
-              task_id, rpc::ErrorType::DEPENDENCY_RESOLUTION_FAILED, &status);
-        }
-      }
-    });
+    io_service_.post(
```

Comment:
passing an io_service_ into CoreWorkerDirectActorTaskSubmitter looks benign since the state is guarded by a mutex; but why do we need this?

Reply:
This PR move io_service_ from CoreWorker to CoreWorkerDirectActorTaskSubmitter, thread model not changed.

Reply:
I think his question is why we need this? I also don’t understand why we need to make this change

Reply:
This change is the main idea of backpressure design, can be referenced in the design doc: https://docs.google.com/document/d/1OGkiOJ-mIxmgNNigO9E7XTCZxzXaj-36u1hQQnAjZgg/edit# . see the `Adjust the sending thread model` part.

Before change: `SubmitActorTask` thread just post the TaskSpec object to the io_service_ and return immediately, the rest of the submitting is in io_service_.
After change: `SubmitActorTask` thread puts TaskSpec object into client queue and then do post, the rest submitting is in io_service_.

![image](https://user-images.githubusercontent.com/4771905/144192719-2993d052-9ec0-4cfc-8574-a8f55f9c5f8a.png)